arch: arc: Add the intial support of memory domain

vonhust opened the pull request:
Refering the ARM's implementation, the initial support of memory
domain in ARC is added:
* changes in MPU drivers
* changes in Kconfig
* codes to configure memory domain during thread swap
* changes in linker script template
* memory domain related macro definitions

the commited codes are simply tested through
samples/mpu/mem_domain_apis_test.

**one change to be noted:**
* gen_kobject_list.py:
For ARC GNU tool chain, the value of DW_AT_location
is DW_FORM_block1, not DW_FORM_exprloc. If not changed,
kobject_hash.gperf cannot be generated correctly. 
Whether this change is corrected, i don't know, as i am not familiar with Dwarf.
appreciate feedback and ideas.

the change in gen_kobject_list.py is removed in this PR.
For ARC  DW_FORM_block1 is generated when the option -gdwarf-2 is used.
If -gdwarf-4 is used, the value of DW_AT_location is DW_FORM_exprloc.


Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: Add the intial support of memory domain

Refering the ARM's implementation, the initial support of memory
domain in ARC is added:
* changes in MPU drivers
* changes in Kconfig
* codes to configure memory domain during thread swap
* changes in linker script template
* memory domain related macro definitions

the commited codes are simply tested through
samples/mpu/mem_domain_apis_test.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -140,6 +143,127 @@ static inline s32_t _mpu_probe(u32_t addr)
+{
+	/*
+	 * The new MPU regions are allocated per type after the statically
+	 * configured regions. The type is one-indexed rather than
```
I think MPUv2 and MPUv3 use a different layout/allocation for the indexes, please update comments. It would help to add a short description of the layout, as well.

vonhust replied:
added

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -140,6 +143,127 @@ static inline s32_t _mpu_probe(u32_t addr)
+}
+
+/**
+ * This internal function check if region is enabled or not
```
checkS

vonhust replied:
fixed

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -140,6 +143,127 @@ static inline s32_t _mpu_probe(u32_t addr)
+	r_size_lshift = (r_size_lshift & 0x3) | ((r_size_lshift >> 7) & 0x1C);
+	r_addr_end = r_addr_start  + (1 << (r_size_lshift + 1));
+
+#elif CONFIG_ARC_MPU_VER == 3
```
for the MPU v3, wouldn't it be more efficient to use the hardware probe to check the region?

vonhust replied:
yes, we can use mpu_probe.

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -140,6 +143,127 @@ static inline s32_t _mpu_probe(u32_t addr)
+
+#if CONFIG_ARC_MPU_VER == 2
+	r_ap = _arc_v2_aux_reg_read(_ARC_V2_MPU_RDP0 + 2 * r_index);
+#elif CONFIG_ARC_MPU_VER == 3
```
see previous remark, hardware probing the address seems more efficient.

vonhust replied:
use mpu_probe as suggested

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -59,7 +59,6 @@ static inline u8_t _get_num_regions(void)
 	return (u8_t)num;
 }
 
-
```
not related to the changes in this PR, but not sure you can/should define a new local variable inside a case block  (u8_t bits in _get_region_attr_by_type below)

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -310,6 +410,149 @@ void arc_core_mpu_region(u32_t index, u32_t base, u32_t size,
+	 * we can stop the iteration immediately once we find the
+	 * matched region that grants permission or denies access.
+	 *
+	 * For ARC MPU v3, overlapping is not supported.
```
For MPU v3, hardware probing would be much more efficient, instead of iterating over all regions.

ruuddw commented on include/arch/arc/arch.h:
```diff
@@ -78,7 +78,128 @@ extern "C" {
+	BUILD_ASSERT_MSG(!(((size) & ((size) - 1))) && (size) >= STACK_ALIGN \
+		 && !((u32_t)(start) & ((size) - 1)), \
+		"the size of the partition must be power of 2" \
+		" and greater than or equal to 2048." \
```
I think the v2 MPU can be (hardware design time) configured for 32byte minimum size in recent ARC cores, instead of the 2k checked for here.

vonhust replied:
change 2048 to the address alignment requirement of MPU

ruuddw commented on include/arch/arc/arch.h:
```diff
@@ -78,7 +78,128 @@ extern "C" {
+
+#ifdef CONFIG_ARC_USERSPACE
+#ifndef _ASMLANGUAGE
+/* Syscall invocation macros. arm-specific machine constraints used to ensure
```
this all is still to be done, but the comments refer to arm-specific stuff -> please change to ARC or remove, not sure the same comments will apply for the ARC implementation

vonhust replied:
fix

ruuddw commented on include/arch/arc/v2/mpu/arc_core_mpu.h:
```diff
@@ -30,6 +30,7 @@ extern "C" {
 /* Thread Stack Region Intent Type */
 #define THREAD_STACK_REGION 0x1
 #define THREAD_STACK_GUARD_REGION 0x2
+#define THREAD_DOMAIN_PARTITION_REGION 0x3
```
above comment mentions Stack regions only, please fix: and make it MPU Region Intent Type. Also, not sure if the THREAD_ prefix matches the domain partion region, memory domains are not thread specific like a stack.

vonhust replied:
fixed

ruuddw commented on scripts/gen_kobject_list.py:
```diff
@@ -441,7 +441,7 @@ def find_kobjects(elf, syms):
 
             loc = die.attributes["DW_AT_location"]
-            if loc.form != "DW_FORM_exprloc":
+            if loc.form != "DW_FORM_exprloc" and loc.form != "DW_FORM_block1":
```
The DW_FORM_block1 might still be used here for compatibility with MetaWare - I'll check with the GNU team.

andrewboie replied:
I think this change should be moved to another patch which is concerned with enabling kernel objects, since everything else here is about enabling the MPU/memory domains.

do DW_FORM_exprloc and DW_FORM_block1 have the same format? subsequent code attempts to parse the data within it to get a memory address.

vonhust replied:
It needs the check from ARC GNU team. From the the mem_domain_api test, it looks no problem. I will make the change in a separate commit

ruuddw reviewed (commented):
Mostly remarks about the code comments, and I suggest to use hardware probing for v3 MPU (could be done later as an optimization as well).

AdithyaBaglody commented:
Please have a look at PR #5065.
Which requires an additional API _arch_mem_domain_configure.

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -310,6 +419,161 @@ void arc_core_mpu_region(u32_t index, u32_t base, u32_t size,
+		pparts = NULL;
+	}
+#if CONFIG_ARC_MPU_VER == 2
+	for (; region_index >= 0; region_index--) {
```
region_index is declared as unsigned int, this loop doesn't end but crashes for region_index=='-1'

vonhust replied:
thanks, region_index should be declared as signed int

ruuddw commented on arch/arc/core/mpu/arc_mpu.c:
```diff
@@ -337,8 +601,8 @@ static void _arc_mpu_config(void)
-	 * According to ARCv2 ISA, the mpu region with smaller
-	 * index has higher priority. The static background mpu
+	 * According to ARCv2 ISA, the MPU region with smaller
+	 * index has higher priority. The static background MPU
```
The logic below for the v2 MPU region layout looks broken. r_index is initialized to the first default region number, but then initialized to 0 again in the for loop. And, the loop loops over all regions, instead of the configured ones only.

andrewboie commented on arch/arc/Kconfig:
```diff
@@ -41,6 +41,7 @@ menu "ARCv2 Family Options"
 config	CPU_ARCV2
 	bool
 	select		ARCH_HAS_STACK_PROTECTION
+	select ARCH_HAS_USERSPACE if ARC_CORE_MPU
```
Hmm. The problem with exporting CONFIG_ARCH_HAS_USERSPACE is that there is a bunch of userspace test code that will be enabled, but this patch doesn't completely implement userspace for ARC, just the memory domain part. Is the privileged/non-privileged implementation in progress? Might be better incorporate into this patch series, and then only enable ARCH_HAS_USERSPACE in a commit at the end so that the tree remains bisectable.

if the rest of the userspace support is not forthcoming for a while, I would remove this and add this later once everything is in place. otherwise our QA testing will suddenly get a ton of errors.

vonhust replied:
The full user space is still in progress, and will be ready in the end of 2018.01. As the mem_domain_api requires CONFIG_USERSPACE, and CONFIG_USERSPACE requires ARCH_HAS_USERPSACE. So I add it here like ARM did.  To avoid the errors of QA, maybe it's better to remove it now and add it back when everything is ready

andrewboie commented on arch/arc/Kconfig:
```diff
@@ -94,11 +95,19 @@ config ARC_FIRQ
+	  If FIRQ is disabled, the handle of interrupts with highest priority
+	  will be same with other interrupts.
+
+config ARC_USERSPACE
```
You could get rid of this option if you don't need it to 'select' other options.
It only exists on x86, for example, to ensure that certain GDT and TSS options are present.
Having said that, I think you will at least need to select THREAD_STACK_INFO for full userspace support.

vonhust replied:
it makes sensor. Now ARC_USERSPACE is useless, i  will remove it. THREAD_STACK_INFO is enabled by ARC_MPU_OPTION.

andrewboie commented on include/arch/arc/v2/linker.ld:
```diff
@@ -134,41 +154,64 @@ SECTIONS {
 	} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
 
-	SECTION_PROLOGUE(_NOINIT_SECTION_NAME,(NOLOAD),) {
+SECTION_PROLOGUE(_NOINIT_SECTION_NAME,(NOLOAD),) {
```
don't think you want to get rid of the leading tab here

vonhust replied:
typo, i will fix it.

Commit:
arch: arc: Add _arch_mem_domain_configure

add the _arch_mem_domain_configure api to match the PR #5065 to avoid
compile failure in the future

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
board: em_starter_kit: add default config for em7d in version 22

* arc MPUv2 is only available in em7d em starterkit version 22.
  Add a default config for this to simply the config.

* The change of board version in menuconfig, e.g, version 23 to
  v22, will not cause other options depending on board version
  change automatically, users still need to change these options
  manually. This is another reason to need a new default config file.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ruuddw reviewed (approved):
Thanks Wayne, this looks good to me now for both V2 and V3 version of the MPU.
